File: src/chapters/minimization/content/code/golden.rs
// Golden-section search for the minimum of a unimodal f on [a, b].
fn golden_section<F: Fn(f64) -> f64>(f: F, mut a: f64, mut b: f64, tol: f64) -> f64 {
    let g = (5.0_f64.sqrt() - 1.0) / 2.0;
    let (mut c, mut d) = (b - g * (b - a), a + g * (b - a));
    let (mut fc, mut fd) = (f(c), f(d));
    while b - a > tol {
        if fc < fd { b = d; d = c; fd = fc; c = b - g * (b - a); fc = f(c); }
        else { a = c; c = d; fc = fd; d = a + g * (b - a); fd = f(d); }
    }
    (a + b) / 2.0
}
fn main() {
    println!("{}", golden_section(|x: f64| (x - 2.0).powi(2) + 1.0, 0.0, 5.0, 1e-8));
}

File: src/pages/practice/guides/matlab/golden.m
function m = golden_section(f, a, b, tol)
% GOLDEN_SECTION  Minimum of a unimodal f on [a,b] by golden-section search.
    if nargin < 4, tol = 1e-8; end
    g = (sqrt(5) - 1)/2;
    c = b - g*(b-a); d = a + g*(b-a); fc = f(c); fd = f(d);
    while b - a > tol
        if fc < fd
            b = d; d = c; fd = fc; c = b - g*(b-a); fc = f(c);
        else
            a = c; c = d; fc = fd; d = a + g*(b-a); fd = f(d);
        end
    end
    m = (a + b)/2;
end

% --- Demo ---
disp(golden_section(@(x) (x-2)^2 + 1, 0, 5));       % -> 2
